server/scsynth/SC_PortAudio.cpp

```diff
@@ -262,31 +264,199 @@ int SC_PortAudioDriver::PortAudioCallback(const void* input, void* output, unsig
     return paContinue;
 }
 
-void SC_PortAudioDriver::GetPaDeviceFromName(const char* device, int* mInOut) {
+void SC_PortAudioDriver::GetPaDeviceFromName(const char* device, int* mInOut, IOType ioType) {
     const PaDeviceInfo* pdi;
     const PaHostApiInfo* apiInfo;
     char devString[256];
     PaDeviceIndex numDevices = Pa_GetDeviceCount();
-    mInOut[0] = paNoDevice;
-    mInOut[1] = paNoDevice;
+    *mInOut = paNoDevice;
 
-    // This tries to find one or two devices that match the given name (substring)
-    // might cause problems for some names...
     for (int i = 0; i < numDevices; i++) {
         pdi = Pa_GetDeviceInfo(i);
+#ifndef __APPLE__
         apiInfo = Pa_GetHostApiInfo(pdi->hostApi);
         strcpy(devString, apiInfo->name);
         strcat(devString, " : ");
+#endif
         strcat(devString, pdi->name);
-        if (strstr(devString, device)) {
-            if (pdi->maxInputChannels > 0)
-                mInOut[0] = i;
-            if (pdi->maxOutputChannels > 0)
-                mInOut[1] = i;
+        // compare strings, but if the string is not empty
+        if (strstr(devString, device) && device && device[0]) {
+            if (ioType == IOType::Input) {
+                if (pdi->maxInputChannels > 0)
+                    *mInOut = i;
+            } else {
+                if (pdi->maxOutputChannels > 0)
+                    *mInOut = i;
+            }
         }
     }
 }
 
+// this function will select default PA devices if they are not defined
+// it will also try to check for some configuration problems
+// numIns, numOuts and sampleRate are only the requested values, can change later
+PaError SC_PortAudioDriver::CheckPaDevices(int* inDevice, int* outDevice, int numIns, int numOuts, double sampleRate) {
+    // make independent checks whether we are using only input, only output, or both input and outputTouched
+    if (numIns && !numOuts) {
```

Comment:
do you mean something like this? 
```c++
if (numIns) {
  if (numOuts) {
     // we have ins and outs
  } else {
    // we have only ins
} else if (numOuts) {
  // we have only outs
}
```
or something else?

Reply:
i mean that this code should look like this:

```cpp
if (numIns && !numOuts) {
    DoOnlyHaveOneIOType(inputData1, outputData2, inputData3);
} else if (! numIns && numOuts) {
    DoOnlyHaveOneIOType(outputData1, inputData2, outputData3);
} else if ...
```

since both branches do the same thing, but with input/output swapped, you can abstract them into a single function.